src/contracts/middleware/BLSSignatureChecker.sol

```diff
@@ -1,508 +1,218 @@
-            signedTotals.signedStakeFirstQuorum -= localStakeObject.firstQuorumStake;
-            signedTotals.signedStakeSecondQuorum -= localStakeObject.secondQuorumStake;
-            
-            // call to ecAdd
-            // aggregateNonSignerPublicKey = aggregateNonSignerPublicKey + nonSignerPublicKey
-            // (input[0], input[1])        = (input[0], input[1])        + (input[2], input[3])
-            // solium-disable-next-line security/no-inline-assembly
-            assembly {
-                success := staticcall(sub(gas(), 2000), 6, input, 0x80, input, 0x40)
-                // Use "invalid" to make gas estimation work
-                switch success
-                case 0 {
-                    invalid()
+        
+        // initialize memory for the quorumStakeTotals
+        QuorumStakeTotals memory quorumStakeTotals;
+        quorumStakeTotals.totalStakeForQuorum = new uint96[](quorumNumbers.length);
+        quorumStakeTotals.signedStakeForQuorum = new uint96[](quorumNumbers.length);
+        // the pubkeyHashes of the nonSigners
+        bytes32[] memory nonSignerPubkeyHashes = new bytes32[](nonSignerStakesAndSignature.nonSignerPubkeys.length);
+        {
+            // the quorumBitmaps of the nonSigners
+            uint256[] memory nonSignerQuorumBitmaps = new uint256[](nonSignerStakesAndSignature.nonSignerPubkeys.length);
+            {
+                // the bitmap of the quorumNumbers
+                uint256 signingQuorumBitmap = BitmapUtils.bytesArrayToBitmap(quorumNumbers);
+
+                for (uint i = 0; i < nonSignerStakesAndSignature.nonSignerPubkeys.length; i++) {
+                    nonSignerPubkeyHashes[i] = nonSignerStakesAndSignature.nonSignerPubkeys[i].hashG1Point();
+                    
+                    // check that the nonSignerPubkeys are sorted and free of duplicates
+                    if (i != 0) {
+                        require(uint256(nonSignerPubkeyHashes[i]) > uint256(nonSignerPubkeyHashes[i - 1]), "BLSSignatureChecker.checkSignatures: nonSignerPubkeys not sorted");
+                    }
+
+                    nonSignerQuorumBitmaps[i] = 
+                        registryCoordinator.getQuorumBitmapByOperatorIdAtBlockNumberByIndex(
+                            nonSignerPubkeyHashes[i], 
+                            referenceBlockNumber, 
+                            nonSignerStakesAndSignature.nonSignerQuorumBitmapIndices[i]
+                        );
+                    
+                    // subtract the nonSignerPubkey from the running apk to get the apk of all signers
+                    apk = apk.plus(
+                        nonSignerStakesAndSignature.nonSignerPubkeys[i]
+                            .negate()
+                            .scalar_mul_tiny(
+                                BitmapUtils.countNumOnes(nonSignerQuorumBitmaps[i] & signingQuorumBitmap) // we subtract the nonSignerPubkey from each quorum that they are a part of, TODO: 
```

Comment:
there's an empty 'TODO' here, not sure what for.
In general the procedure makes sense, but do we have a doc that abstracts away from the code, so that e.g. someone who is just good at cryptography could easily review?